benchmarks/persistence/persistence.py

```diff
@@ -0,0 +1,229 @@
+        current_branch = repo.active_branch.name
+        S3_PATH_FORBIDDEN_CHARACTERS = ["/", "\\"]
+        sanitized_branch = ""
+        for char in S3_PATH_FORBIDDEN_CHARACTERS:
+            sanitized_branch = current_branch.replace(char, "-")
+        return sanitized_branch
+
+    @_workflow_run_identifier.default
+    def _default_workflow_id(self) -> str:
+        if "GITHUB_RUN_ID" not in os.environ:
+            raise ValueError("The environment variable GITHUB_RUN_ID is not set.")
+        return os.environ["GITHUB_RUN_ID"]
+
+    def get_full_file_path(self) -> str:
+        """Construct the path of a result object.
+
+        Returns:
+            The path of the result object.
+        """
+        experiment_identifier = self._benchmark.name
+        metadata = self._result.metadata
+        file_usable_date = metadata.start_datetime.strftime("%Y-%m-%d")
+        bucket_path_key = (
+            f"{experiment_identifier}/{self._branch}/{metadata.latest_baybe_tag}/"
+            + f"{file_usable_date}/"
+            + f"{metadata.commit_hash}/{self._workflow_run_identifier}/"
+            + "result.json"
+        )
+        return bucket_path_key
+
+    def get_object(self) -> dict:
+        """Construct the object to be persisted.
+
+        Returns:
+            The object to be persisted.
+        """
+        return self._result.to_dict() | self._benchmark.to_dict()
+
+
+@define
+class S3PersistenceHandler:
+    """Class for persisting experiment results in an S3 bucket."""
+
+    _bucket_name: str = field(validator=instance_of(str), init=False)
+    """The name of the S3 bucket where the results are stored."""
+
+    _object_session = boto3.session.Session()
+    """The boto3 session object. This will load the respective credentials
+    from the environment variables within the container."""
+
+    def write_object(self, object: PersistenceObjectInterface) -> None:
+        """Store the result of a benchmark.
+
+        This method will store the result of a benchmark in an S3 bucket.
+        The S3-key of the Java Script Notation Object will be the experiment identifier,
+        the branch, the BayBE-version, the start datetime, the commit hash and the
+        workflow run identifier.
+
+        Args:
+            object: The object to be persisted.
+        """
+        client = self._object_session.client("s3")
+
+        client.put_object(
+            Bucket=self._bucket_name,
+            Key=object.get_full_file_path(),
+            Body=json.dumps(object.get_object()),
+            ContentType="application/json",
+        )
+
+
+@define
+class LocalFileSystemObjectCreator:
+    """Class for creating the path of a result object."""
+
+    _benchmark: Benchmark = field(validator=instance_of(Benchmark))
+    """The benchmark for which the result is stored."""
+
+    _result: Result = field(validator=instance_of(Result))
+    """The result of the benchmark."""
+
+    _path: str = field(validator=instance_of(str))
```

Comment:
Any reason why these are `str` and not `Path` objects? As far as I know, those are significantly more stable and better to use, specifically on different OS.

Reply:
This is also potentially causing problems in other parts, as you e.g. explicitly use "/" as a separator which does not work on all OS, so the logic here should be changed in general to `Path` obejcts, ideally while still accepting `str` (and `Path`) as inputs.